Add test guideline when contributing code (#1052)
mattermost/mattermost-developer-documentation: site/content/contribute/getting-started/test-policy.md

---

> title: "Test Policy"

Reviewer: How do we all feel about the term "policy" vs "guideline"? It may be minor and I'm definitely not 5/5 on avoiding the term policy when we can, but I find the term "guidelines" preferable.

Author: The original ask was to have a "unit test policy" but I honestly don't think I was able to deliver such expectation. It even expanded (in a different direction) into a broader scope which included end-to-end test.
I recently mentioned this uncertainty during our 1:1 but I'm not sure what's the best term to use given the content of the document.

Author: Per offline discussion, it's agreed to change from policy to guideline.

Now updated.

This document was written to stress the importance of tests for every pull request being submitted. This is not to meet higher code coverage but rather to write effective and well-planned use cases depending on the changes being made.

This document aims to serve as the foundation of our test policy and will serve as a guide on why we don't merge code without tests. But of course, there's always an exception. If it is not possible for some reason, let the reviewers know by writing a description to open up a discussion and to fully understand the situation you are facing.

Test categories
---------------
1. __Unit test__ - verify that individual, isolated parts work as expected.
2. __Integration test__ - verify that several units work together in harmony.
3. __End-to-end test__ - exercises most of the parts of a large application.

Note: Wisdom and definition mostly taken from (a) https://martinfowler.com/testing/ and (b) https://kentcdodds.com/.

Not all test types are required in a single pull request. Only write whichever test types are most effective and appropriate.

In general, when to have these tests?
-------------------------------------

> 1. For all files written in the main language/s of the repo; could be Javascript/Typescript including JSX/TSX, Go, or both which are exported such as functions, modules or components being used in several places.

Author: Done, thanks!

2. Unexported function or method, which has low or no test coverage from the parent exported function/method, that affects critical functionality or behavior of the application.
3. New features and bug fixes, most especially customer and community bugs.

And when is it fine not to have these tests?
--------------------------------------------

> 1. For implementation detail of standard library or external package that is implicitly covered by the standard library or external package itself.

Author: Done, thanks!

> 1. Implementation detail of standard library or external package which are implicitly covered by the standard library or external package itself.

Reviewer 2: ```suggestion
1. For implementation detail of standard library or external package that is implicitly covered by the standard library or external package itself.
```
